Multicursor Implementation

adamharrison opened the pull request:
Rewrote things to be compatible with a multicursor implementation.

**Major Features**
- Should be fully backwards compatible with all plugins (i.e. if you're using a single cursor, nothing changes). If you have a multicursor, could have some weird stuff happening, but *should* have things occur to the first cursor.
- All translations, and newlines, and text input and stuff can now be targeted per-cursor; by default, they target all cursors if prefixed with an nil; otherwise, they take a cursor index.
- Multiple selections possible.
- Multiple clipboard buffers; one per cursor are used. System clipboard contains a copy separated by the line separator.
- Holding control, and dragging will create multiple cursors.
- All in all, only about 60 SLOC added.

There are likely *many* bugs. This is a first iteration; but I thought I'd throw up the pull request so people could leave comments, request features, and try the thing out if they want.

I am not super familiar with using multicursors in for day-to-day usage, so would heavily appreciate feature requests!

adamharrison commented:
To close an issue; should fix #96 .

franko commented:
Wow, kudos for tackling this feature, I never had the courage to start working on that but this is a super powerful feature to have.

adamharrison commented:
If no one has any complaints/requests, I'll merge this on 2021-06-11 after doing a final bug pass.

jgmdev commented:
Pretty sweet!, was testing and seems like some plugins like autocomplete will need some modification to insert the text on the multiple locations. Bindings that are used in other editors like vscode or geany are ctrl+shift+up or alt+shift+up to add cursors above and ctrl+shift+down or alt+shift+down to add cursors below.

jgmdev commented:
Something I notice with autoinsert plugin:

1. doc:move_to(-1) doesn't seems to work so the plugin puts the cursor as shown `()|` instead of `(|)` inside the parenthesis
2. doc:delete_to(1) throws an error as shown in screenshot below:
![2021-06-08_03:08:41](https://user-images.githubusercontent.com/1702572/121139603-edce2080-c806-11eb-86da-9ec5586d1d12.png)

Lastly this snippet that is in charge of inserting a pair of characters around a selection is putting the cursor at the beginning of the selection after inserting the pair instead of at the end, for example, if I write `my text`, select it and type `"` I get `|"my text"` instead of `"my text"|`, here is the snippet from autoinsert plugin:

```lua
if mapping and self.doc:has_selection() then
  local l1, c1, l2, c2, swap = self.doc:get_selection(true)
  self.doc:insert(l2, c2, mapping) -- the closing char which is "
  self.doc:insert(l1, c1, text) -- user typed char which is also "
  self.doc:set_selection(l1, c1, l2, c2 + 2) -- should move cursor at end of selection
  return
end
```

adamharrison commented:
> Pretty sweet!, was testing and seems like some plugins like autocomplete will need some modification to insert the text on the multiple locations. Bindings that are used in other editors like vscode or geany are ctrl+shift+up or alt+shift+up to add cursors above and ctrl+shift+down or alt+shift+down to add cursors below.

Perfect, I'll add those in as commands.

> doc:move_to(-1) doesn't seems to work so the plugin puts the cursor as shown ()| instead of (|) inside the parenthesis
> doc:delete_to(1) throws an error as shown in screenshot below:

Yeah. These require a nil in front now; but it's probably a better to just make three extra functions `move_to_cursor` or something, that take the new index, and leave these ones alone. I'll do that.

> Lastly this snippet that is in charge of inserting a pair of characters around a selection is putting the cursor at the beginning of the selection after inserting the pair instead of at the end, for example, if I write my text, select it and type " I get |"my text" instead of "my text"|, here is the snippet from autoinsert plugin:

That is odd. I'll investigate this.

ChildishGiant commented:
The bind I use in vscode for this is clicking somewhere when holding ctrl to place a new cursor there

adamharrison commented:
> The bind I use in vscode for this is clicking somewhere when holding ctrl to place a new cursor there

Yeah, this is currently the case. This one's not rebindable via the keymap (may have to implement a new system to deal with mouse actions), but it's good to hear my vague memory that someone had once mentioned that being the shortcut was accurate!

adamharrison commented:
In VS code, what happens when cursors double-up? Do they merge? Or do you emit n characters per keystroke, where n is the amount of overlapping cursors?

Also, for cntrl+shift+up, do *all* cursors split upwards, or only the cursor on the lowest line?

adamharrison commented:
> Pretty sweet!, was testing and seems like some plugins like autocomplete will need some modification to insert the text on the multiple locations. Bindings that are used in other editors like vscode or geany are ctrl+shift+up or alt+shift+up to add cursors above and ctrl+shift+down or alt+shift+down to add cursors below.

This has been done; assuming that we only create cursors above a cursor if there isn't already a cursor there.

> doc:move_to(-1) doesn't seems to work so the plugin puts the cursor as shown ()| instead of (|) inside the parenthesis
> doc:delete_to(1) throws an error as shown in screenshot below:

These have been reverted; doc:move_to_cursor is now used internally; but externally, doc:move_to will move all cursors.

> Lastly this snippet that is in charge of inserting a pair of characters around a selection is putting the cursor at the beginning of the selection after inserting the pair instead of at the end, for example, if I write my text, select it and type " I get |"my text" instead of "my text"|, here is the snippet from autoinsert plugin:

I removed swap; but I've added it back for the legacy functions.

> The bind I use in vscode for this is clicking somewhere when holding ctrl to place a new cursor there

I think I'll make a separate pull request for adding in mouse commands for keymap.

jgmdev commented:
> This has been done; assuming that we only create cursors above a cursor if there isn't already a cursor there.

Tested and ctrl + shift + up works but ctrl + shift + down does not

> These have been reverted; doc:move_to_cursor is now used internally; but externally, doc:move_to will move all cursors.

Nice! I can confirm that the autoinsert plugin is now working as before 😃 

> I removed swap from `get_selection`; but I've just added it back for the legacy functions.

Works!

ChildishGiant commented:
> In VS code, what happens when cursors double-up? Do they merge? Or do you emit n characters per keystroke, where n is the amount of overlapping cursors?

They just merge into one cursor, one output per key stroke. As for the other questions I don't know how to do the splitting up and down in vscode (the binds here don't work) so I haven't been able to check

adamharrison commented:
> Tested and ctrl + shift + up works but ctrl + shift + down does not

Yeah, sorry about that. Fixed.

> They just merge into one cursor, one output per key stroke. As for the other questions I don't know how to do the splitting up and down in vscode (the binds here don't work) so I haven't been able to check

OK. I've made it so that cursors should merge after moving.

jgmdev commented:
Just noticed that the changes on Multicursor branch cause lite to go nuts when pressing ctrl+x to cut some text in the editor, it freezes and cpu usage rises up.

adamharrison commented:
Ah. Yeah; sorry, I used the non-cursor-function in a selection loop.

adamharrison commented:
OK, I'm going to do a thorough bug-testing pass now, and leave it up for a day or two more to get more comments, but otherwise, pending that pass, and final review from Franko, I think this should be merged.

adamharrison commented:
I'll actually just leave this as a draft until I do the full bug pass. There are still bugs here.

franko commented:
@adamharrison thank you for this amazing work. I think what you are trying to implement is very difficult to integrate with the application since it was all designed only for a single cursor.

Take your time to work on that, there is no hurry. I am sure at the end this is going to be an amazing feature.

adamharrison commented:
OK, I think I squashed a few more bugs. I think I've got most of them. I think *now*, it's probably good for an external review.

franko commented:
Great, thank you Adam for this amazing work.

I have tested and it looks fine to me but the problem is that I am not a multi-cursor connoisseur so the depth of my testing didn't go very far. I will try to test more this feature but I am favorable to merge this PR right away or may be we can give some time to other people to test.

For my own education to the multi-cursor usage I made some research and I found this interesting introduction:

https://dev.to/codevault/multi-cursor---visual-studio-code-productivity-features---part-1-1gha

then comparing to our implementation I thought we are missing the "Add multi-cursor to next occurrence" feature that is bound to ctrl+d in vscode. May be we are missing also the "Add multi-cursor to all occurrences".

I am wondering if, now that we have multi-cursor we shouldn't complement it with the additional functionalities I mentioned above to make it really complete.

I don't mean we should delay the merge but maybe we should put some thoughts to this for some further development to complete the feature. Now that the core mechanism is there we may ensure that all the additional commands and keybindings are there too to use multi-cursors to their full potential.

I am also favorable to add a mechanism to support the ctrl+click, I may even contribute an implementation if you want.

adamharrison commented:
I think those'd be good. I think we should just merge this (maybe after someone who's a bit more familiar with multicursor reviews; @jgmdev or @ChildishGiant ) so it doesn't get too big and unwieldly (as it feels useful as-is); and we can incrementally add in extra functionality to improve the usability as we go on.

ChildishGiant commented:
I'll try to test a build at some point, my os installs have been playing up recently and building C isn't my usual wheelhouse so it'll take me a bit

redtide commented:
> I'll try to test a build at some point, my os installs have been playing up recently and building C isn't my usual wheelhouse so it'll take me a bit

I wonder why CI is not working on each commit, so that could be possible to have also temporary builds to download (artifacts)

franko commented:
> I wonder why CI is not working on each commit, so that could be possible to have also temporary builds to download (artifacts)

I disabled CI on commits so CI doesn't run anymore except on demand. I thought it was wasteful and not really useful. We don' have window's builds and the macOS application may be not correctly packaged (not sure about that).

On the other side we may add some tags for developer previews or betas and provide the binaries. We are just not ready to build them automatically using CI. I may provide the binaries myself like I am doing for ordinary releases. Let's just coordinate to produce some pre-release tags.

jgmdev commented:
This has been getting better and better! I found an issue with the autoinsert plugin when pressing backspace, lets say I have this file:

```json
{
 "test": "content"
}
```
If i place the cursor on beginning of second line `|"test": "content"` and press backspace to get something like:

```json
{"test": "content"
}
```
Lite freezes and will not respond.

This is the autoinsert plugin code that is causing the issue and seems to be caused by the delete_to(1) call:

```lua
["autoinsert:backspace"] = function()
    local doc = core.active_view.doc
    local l, c = doc:get_selection()
    local chr = doc:get_char(l, c)
    if config.autoinsert_map[doc:get_char(l, c - 1)] and is_closer(chr) then
      doc:delete_to(1)
    end
    command.perform "doc:backspace"
  end,
```

adamharrison commented:
Ah yes; my bad. Forgot to change that one. I'm kinda surprised I didn't see it, to be honest. Fixed.

adamharrison commented:
Just a heads up to anyone interested, I'm going to aim to merge this on the 18th; I'll do one last review on that date before merging.

jgmdev commented:
Has been working stable after several hours of code session.

adamharrison commented:
Oooh, I like it. I'll see what I can do.

adamharrison commented:
OK, that should be in.

adamharrison commented:
OK. I did another pass and it looks good. Fingers crossed; merging in!

franko commented:
Kudos for this major achievement!
